### ISSUE TITLE
Null Reference Exception coming from APIGatewayProxyFunction

### ISSUE THREAD

**Issue by dsartain18** [aws/aws-lambda-dotnet#168]
I've beaten this one to death as best I can before posting here because I've got to be missing something, but at this point I just can't figure it out. 

I've published to AWS Lambda from Visual Studio 2017 using the serverless template and S3 stack as well as just publishing straight there. I've also tried just uploading a zip file. Every time I test from the API Gateway or the tools in Visual Studio, I get the following exception.

```
{
	  "errorType": "AggregateException",
	  "errorMessage": "One or more errors occurred. (Object reference not set to an instance of an object.)",
	  "stackTrace": [
		"at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)",
		"at System.Threading.Tasks.Task`1.GetResultCore(Boolean waitCompletionNotification)",
		"at lambda_method(Closure , Stream , Stream , ContextInfo )"
	  ],
	  "cause": {
		"errorType": "NullReferenceException",
		"errorMessage": "Object reference not set to an instance of an object.",
		"stackTrace": [
		  "at Amazon.Lambda.AspNetCoreServer.APIGatewayProxyFunction.MarshallRequest(InvokeFeatures features, APIGatewayProxyRequest apiGatewayRequest)",
		  "at Amazon.Lambda.AspNetCoreServer.APIGatewayProxyFunction.<FunctionHandlerAsync>d__12.MoveNext()"
		]
	  }
	}
```

I used the following blog to make sure I deployed correctly:
https://aws.amazon.com/blogs/developer/deploy-an-existing-asp-net-core-web-api-to-aws-lambda/

Lambda Entry Point:
```
public class LambdaEntryPoint : Amazon.Lambda.AspNetCoreServer.APIGatewayProxyFunction
    {
        protected override void Init(IWebHostBuilder builder)
        {
            try
            {
                builder
                    .UseContentRoot(Directory.GetCurrentDirectory())
                    .UseStartup<Startup>()
                    .UseApiGateway();
            }
            catch (Exception ex)
            {
                LambdaLogger.Log("Exception throw in LambdaEntryPoint: " + ex);
            }
        }
    }
```

.csproj contents
```
<Project Sdk="Microsoft.NET.Sdk.Web">

  <PropertyGroup>
    <TargetFramework>netcoreapp1.0</TargetFramework>
    <PackageTargetFallback>$(PackageTargetFallback);portable-net45+win8+wp8+wpa81;</PackageTargetFallback>
    <PreserveCompilationContext>false</PreserveCompilationContext>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Amazon.Lambda.AspNetCoreServer" Version="0.10.2-preview1" />
    <PackageReference Include="Amazon.Lambda.Core" Version="1.0.0" />
    <PackageReference Include="Amazon.Lambda.Logging.AspNetCore" Version="1.0.0" />
    <PackageReference Include="Amazon.Lambda.Tools" Version="1.8.0" />
    <PackageReference Include="AWSSDK.Core" Version="3.3.17.10" />
    <PackageReference Include="AWSSDK.Extensions.NETCore.Setup" Version="3.3.3" />
    <PackageReference Include="Microsoft.AspNetCore" Version="1.0.5" />
    <PackageReference Include="Microsoft.AspNetCore.Mvc" Version="1.0.4" />
    <PackageReference Include="Microsoft.AspNetCore.StaticFiles" Version="1.0.3" />
    <PackageReference Include="Microsoft.EntityFrameworkCore" Version="1.0.5" />
    <PackageReference Include="Microsoft.EntityFrameworkCore.Relational" Version="1.0.5" />
    <PackageReference Include="Microsoft.EntityFrameworkCore.Relational.Design" Version="1.0.5" />
    <PackageReference Include="Microsoft.EntityFrameworkCore.SqlServer" Version="1.0.5" />
    <PackageReference Include="Microsoft.EntityFrameworkCore.SqlServer.Design" Version="1.0.5" />
    <PackageReference Include="Microsoft.EntityFrameworkCore.Tools" Version="1.0.2" />
    <PackageReference Include="Microsoft.Extensions.Logging.Debug" Version="1.0.2" />
    <PackageReference Include="Microsoft.VisualStudio.Web.BrowserLink" Version="1.0.1" />
    <PackageReference Include="Microsoft.VisualStudio.Web.CodeGeneration.Design" Version="1.0.0" />
  </ItemGroup>

  <ItemGroup>
    <DotNetCliToolReference Include="Microsoft.EntityFrameworkCore.Tools.DotNet" Version="1.0.2" />
    <DotNetCliToolReference Include="Amazon.Lambda.Tools" Version="1.8.0" />
    <DotNetCliToolReference Include="Microsoft.VisualStudio.Web.CodeGeneration.Tools" Version="1.0.0" />
  </ItemGroup>

  <ItemGroup>
    <Folder Include="Controllers\" />
    <Folder Include="Data\" />
    <Folder Include="Models\" />
    <Folder Include="Views\Shared\" />
    <Folder Include="wwwroot\css\" />
    <Folder Include="wwwroot\js\" />
  </ItemGroup>

</Project>
```

I tried getting the serverless invoke local command to work, but it wouldn't for whatever reason. So, I'm at a loss on how to debug this and get to what's throwing the null reference exception out of the SDK.

Thanks, and please let me know if you need anything else from me.

**Comment by normj** [comment]
To help debug this can you add the environment variable **LAMBDA_NET_SERIALIZER_DEBUG** with a value of **true** to the Lambda function.

This will cause the raw JSON coming into the Lambda function be written into CloudWatch logs for the Lambda function. Once you post the RAW JSON I'll try and debug what is going on.

![image](https://user-images.githubusercontent.com/1653751/30979404-68114928-a432-11e7-9f2b-abf10a3ce94c.png)

**Comment by dsartain18** [comment]
@normj The entire response of that piece of the log is:

```
Lambda Deserialize Amazon.Lambda.APIGatewayEvents.APIGatewayProxyRequest:
{}
```
![cw](https://user-images.githubusercontent.com/15874631/30983061-ec6b1138-a44e-11e7-8abc-0c482411fdaa.PNG)

**Comment by dsartain18** [comment]
Adding screen shots of test setup in case that helps.

![config](https://user-images.githubusercontent.com/15874631/30983283-9bda6682-a44f-11e7-8fc9-d429ff1fb531.PNG)

![es](https://user-images.githubusercontent.com/15874631/30983284-9bde8d7a-a44f-11e7-8068-f7d6db7d0f6f.PNG)

![test](https://user-images.githubusercontent.com/15874631/30983331-bd93ab26-a44f-11e7-81c8-907365633496.PNG)

**Comment by dsartain18** [comment]
@normj I tried it again after looking at the Fiddler request from the API Gateway 

Request:
`{"pathWithQueryString":"/api/users","headers":{},"stageVariables":{},"withAuthorization":false}`

Log:
```
Lambda Deserialize Amazon.Lambda.APIGatewayEvents.APIGatewayProxyRequest:
{
    "pathWithQueryString": "/api/users",
    "headers": {},
    "stageVariables": {},
    "withAuthorization": false
}
```

In the API Gateway, I do have the Authorization in the Method Request set to AWS_IAM.

I hope that helps.

**Comment by normj** [comment]
Amazon.Lambda.AspNetCoreServer Assumes the request coming in from API Gateway is a proxy request, http://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-set-up-simple-proxy.html, which is what the default serverless.template will configure. The request you are sending seem to short to be a proxy request because it is missing key parts like the HTTP Method and resource path. When testing in Visual Studio in the **Example Requests** drop down box select **API Gateway AWS Proxy** example and then modify the path and httpMethod to what you are using.

**Comment by dsartain18** [comment]
I got the template and I think I made the necessary modifications, but now I get a 404 error from the VS tester, and a 500 from the API Gateway using the Proxy function. 

```
{
  "body": "{\"test\":\"body\"}",
  "resource": "/{proxy+}",
  "requestContext": {
    "resourceId": "123456",
    "apiId": "1234567890",
    "resourcePath": "/{proxy+}",
    "httpMethod": "GET",
    "requestId": "c6af9ac6-7b61-11e6-9a41-93e8deadbeef",
    "accountId": "123456789012",
    "identity": {
      "apiKey": null,
      "userArn": null,
      "cognitoAuthenticationType": null,
      "caller": null,
      "userAgent": "Custom User Agent String",
      "user": null,
      "cognitoIdentityPoolId": null,
      "cognitoIdentityId": null,
      "cognitoAuthenticationProvider": null,
      "sourceIp": "127.0.0.1",
      "accountId": null
    },
    "stage": "prod"
  },
  "queryStringParameters": {
    "foo": "bar"
  },
  "headers": {
    "Via": "1.1 08f323deadbeefa7af34d5feb414ce27.cloudfront.net (CloudFront)",
    "Accept-Language": "en-US,en;q=0.8",
    "CloudFront-Is-Desktop-Viewer": "true",
    "CloudFront-Is-SmartTV-Viewer": "false",
    "CloudFront-Is-Mobile-Viewer": "false",
    "X-Forwarded-For": "127.0.0.1, 127.0.0.2",
    "CloudFront-Viewer-Country": "US",
    "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8",
    "Upgrade-Insecure-Requests": "1",
    "X-Forwarded-Port": "443",
    "Host": "1234567890.execute-api.us-east-1.amazonaws.com",
    "X-Forwarded-Proto": "https",
    "X-Amz-Cf-Id": "cDehVQoZnx43VYQb9j2-nvCh-9z396Uhbp027Y2JvkCPNLmGJHqlaA==",
    "CloudFront-Is-Tablet-Viewer": "false",
    "Cache-Control": "max-age=0",
    "User-Agent": "Custom User Agent String",
    "CloudFront-Forwarded-Proto": "https",
    "Accept-Encoding": "gzip, deflate, sdch"
  },
  "pathParameters": {
    "proxy": "/api/Users"
  },
  "httpMethod": "GET",
  "stageVariables": {
    "baz": "qux"
  },
  "path": "/api/Users"
}
```

My controller attribute routing is:
```
[Produces("application/json")]
[Route("api/Users")]
```

Did I configure the path and http method incorrectly?

**Comment by codemaster** [comment]
I'm also receiving this error:

```{
    "errorType": "AggregateException",
    "errorMessage": "One or more errors occurred. (Object reference not set to an instance of an object.)",
    "stackTrace": [
        "at System.Threading.Tasks.Task`1.GetResultCore(Boolean waitCompletionNotification)",
        "at lambda_method(Closure , Stream , Stream , LambdaContextInternal )"
    ],
    "cause": {
        "errorType": "NullReferenceException",
        "errorMessage": "Object reference not set to an instance of an object.",
        "stackTrace": [
            "at Amazon.Lambda.AspNetCoreServer.APIGatewayProxyFunction.MarshallRequest(InvokeFeatures features, APIGatewayProxyRequest apiGatewayRequest, ILambdaContext lambdaContext)",
            "at Amazon.Lambda.AspNetCoreServer.APIGatewayProxyFunction.<FunctionHandlerAsync>d__13.MoveNext()"
        ]
    }
}```

I have tried enabling "Use Lambda Proxy integration", but there has been zero change with this option enabled.

**Comment by danielfmsouza** [comment]
I'm also having the exact same problem reported above by @codemaster .

I compiled using .NET Core 2 (2.0.5).

Any updates on this?

**Comment by dsartain18** [comment]
@codemaster @danielfmsouza I haven't made any progress on it because it seemed that the problem was more with my version of the Serverless template. I didn't have the time to spend learning all of that yet. As it didn't seem to be code related, I had to stop working on it for the time being.

**Comment by normj** [comment]
@danielfmsouza Can you provide a sample request to help us diagnose what is going on? I [mentioned up above](https://github.com/aws/aws-lambda-dotnet/issues/168#issuecomment-332897736) how you can enable an environment variable to capture the original JSON request to the CloudWatch Logs.

**Comment by inabeel** [comment]
I have similar problem. I am trying to follow Blog API template from lambda. But when I post object through Postman. I get Null object in APIGatewayProxyRequest request's Body.

Secondly, It also throws exception when try to run Blog API code as it is "Must have one hash key defined for the table TableName\n   at Amazon.DynamoDBv

**Comment by danielfmsouza** [comment]
@normj I've tested like you mentioned, and see what is happening:

1. When I try with a full proxy request from VS or just executing via Postman, I always get empty 200's responses (when it was supposed to return data). See the example below:
Request:
```
{
  
  "resource": "/{proxy+}",
  "requestContext": {
    "resourceId": "123456",
    "apiId": "1234567890",
    "resourcePath": "/{proxy+}",
    "httpMethod": "GET",
    "requestId": "c6af9ac6-7b61-11e6-9a41-93e8deadbeef",
    "accountId": "12345678912",
    "identity": {
      "apiKey": null,
      "userArn": null,
      "cognitoAuthenticationType": null,
      "caller": null,
      "userAgent": "Custom User Agent String",
      "user": null,
      "cognitoIdentityPoolId": null,
      "cognitoIdentityId": null,
      "cognitoAuthenticationProvider": null,
      "sourceIp": "127.0.0.1",
      "accountId": null
    },
    "stage": "prod"
  },
  "queryStringParameters": {
   },
  "headers": {
    "Via": "1.1 08f323deadbeefa7af34d5feb414ce27.cloudfront.net (CloudFront)",
    "Accept-Language": "en-US,en;q=0.8",
    "CloudFront-Is-Desktop-Viewer": "true",
    "CloudFront-Is-SmartTV-Viewer": "false",
    "CloudFront-Is-Mobile-Viewer": "false",
    "X-Forwarded-For": "127.0.0.1, 127.0.0.2",
    "CloudFront-Viewer-Country": "US",
    "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8",
    "Upgrade-Insecure-Requests": "1",
    "X-Forwarded-Port": "443",
    "Host": "1234567890.execute-api.us-east-1.amazonaws.com",
    "X-Forwarded-Proto": "https",
    "X-Amz-Cf-Id": "cDehVQoZnx43VYQb9j2-nvCh-9z396Uhbp027Y2JvkCPNLmGJHqlaA==",
    "CloudFront-Is-Tablet-Viewer": "false",
    "Cache-Control": "max-age=0",
    "User-Agent": "Custom User Agent String",
    "CloudFront-Forwarded-Proto": "https",
    "Accept-Encoding": "gzip, deflate, sdch"
  },
  "pathParameters": {
    "proxy": "api/v1/geos/1"
  },
  "httpMethod": "GET",
  "stageVariables": {
    "baz": "qux"
  },
  "path": "api/v1/geos/1"
}
```
Response:
`{"statusCode":200,"headers":{"Content-Length":"0"},"body":"","isBase64Encoded":false}`

2. When calling from a ScheduledExecution, I always receive the error mentioned previously by @codemaster:
CF template property (It was working fine before migration, the only change made to template was the Runtime set as "dotnetcore2.0"):

```
		"ScheduledExecution": {
		  "Type": "AWS::Events::Rule",
		  "Properties": {
			"Description": "Scheduled execution of the Lambda function",
			"ScheduleExpression": "rate(1 minute)",
			"State": "ENABLED",
			"Targets": [
			  {
				"Arn": { "Fn::GetAtt" : [ "BgGeos", "Arn" ] },
				"Id": "BgGeos",
				"Input": "{  \"headers\": {    \"Accept\": \"application/json\",    \"Host\": \"1234567890.execute-api.us-east-1.amazonaws.com\"  },  \"pathParameters\": {    \"proxy\": \"/api/v1/geos/1\"  },  \"httpMethod\": \"GET\",  \"path\": \"/api/v1/geos/1\"}"
				}
			]
		  }
		}
```

Error (from CloudWatch):
```
One or more errors occurred. (Object reference not set to an instance of an object.): AggregateException
at System.Threading.Tasks.Task`1.GetResultCore(Boolean waitCompletionNotification)
at lambda_method(Closure , Stream , Stream , LambdaContextInternal )

at Amazon.Lambda.AspNetCoreServer.APIGatewayProxyFunction.MarshallRequest(InvokeFeatures features, APIGatewayProxyRequest apiGatewayRequest, ILambdaContext lambdaContext)
at Amazon.Lambda.AspNetCoreServer.APIGatewayProxyFunction.<FunctionHandlerAsync>d__13.MoveNext()

```

I was wondering if the error mentioned above could being suppressed somehow (on the first case) and return a 200 empty response instead...

Thanks!

**Comment by normj** [comment]
I just released version 2.0.1 of `Amazon.Lambda.AspNetCoreServer` that has some new defensive code in there for marshalling that should help this issue. Please update your dependency and retry.

**Comment by codemaster** [comment]
Thank you, @normj! I updated my `Amazon.Lambda.AspNetCoreServer` package to 2.0.1, kicked off a dotnet restore/build/publish and reuploaded my Lambda function. Sadly, I still encounter the error:

```{
  "errorType": "AggregateException",
  "errorMessage": "One or more errors occurred. (Object reference not set to an instance of an object.)",
  "stackTrace": [
    "at System.Threading.Tasks.Task`1.GetResultCore(Boolean waitCompletionNotification)",
    "at lambda_method(Closure , Stream , Stream , LambdaContextInternal )"
  ],
  "cause": {
    "errorType": "NullReferenceException",
    "errorMessage": "Object reference not set to an instance of an object.",
    "stackTrace": [
      "at Amazon.Lambda.AspNetCoreServer.APIGatewayProxyFunction.MarshallRequest(InvokeFeatures features, APIGatewayProxyRequest apiGatewayRequest, ILambdaContext lambdaContext)",
      "at Amazon.Lambda.AspNetCoreServer.APIGatewayProxyFunction.<FunctionHandlerAsync>d__13.MoveNext()"
    ]
  }
}```

**Comment by danielfmsouza** [comment]
@normj Thank you! I've updated the packages (also Lambda.Tools) and it's working fine now!

Thanks!

**Comment by codemaster** [comment]
@normj Any update on this? Is there anything I can do to provide additional information on this issue? 

If it helps, here are the packages that my project is utilizing:

```
  <ItemGroup>
    <PackageReference Include="Microsoft.AspNetCore.All" Version="2.0.3" />
    <PackageReference Include="Google.Apis.Sheets.v4" Version="1.32.0.1084" />
    <PackageReference Include="Autofac" Version="4.6.2" />
    <PackageReference Include="Autofac.Extensions.DependencyInjection" Version="4.2.0" />
    <PackageReference Include="Amazon.Lambda.AspNetCoreServer" Version="2.0.1" />
  </ItemGroup>
```

**Comment by codemaster** [comment]
I was able to figure out what was going on: I needed to have an API Gateway trigger the Lambda function with the proper URI. Triggering the lambda via a test payload wouldn't work because the lambda test functionality has no knowledge of a URI, hence the error.

**Comment by normj** [comment]
Closing for lack of activity and it sounds like the issue was resolved. As this issue has gotten long and hard to follow. If there are still issues please open a new issue.

**Comment by jpnolin** [comment]
Quick comment if you are here maybe searching for why you're API Gateway cannot contact the lambda or you get a 404 code when calling the lambda. If like me you are using the barebone lambda example, dont forget that there is a "api" part in the path of the web API. For example in the ValueController, you should call api/values not just values. Very stupid but took me an hour to find sadly.

**Comment by jamesrwaugh** [comment]
I had this issue, but as pointed out, it is intended to be used for lamda proxy integrations. The following solved it for me.

* When you go to Actions -> Create Resource, I found **Configure as proxy resource** must be checked, 
* Afterwards, "**Use Lambda Proxy Integration**" must also be checked in the "Integration Request" page of the API Gateway config.

![Screenshot from 2019-09-26 12-07-08](https://user-images.githubusercontent.com/1207380/65705487-8b46e800-e056-11e9-825a-85ea56ceadb6.png)

**Comment by Eric-ITaaS** [comment]
Can that be set after the function is created, I can't find that option in the console.

**Comment by diegosasw** [comment]
@jamesrwaugh selecting Use Lambda Proxy integration worked for me. It's the first time I'm playing with Lambda and I've created a very simple lambda for aspnetcore3.1 with just a middleware that responds "Hello World" when GET /

I have been struggling for hours when testing from a REST API Gateway with a GET / resource with the following exception
```
Object reference not set to an instance of an object.: NullReferenceException
   at Amazon.Lambda.AspNetCoreServer.APIGatewayProxyFunction.MarshallRequest(InvokeFeatures features, APIGatewayProxyRequest apiGatewayRequest, ILambdaContext lambdaContext)
   at Amazon.Lambda.AspNetCoreServer.AbstractAspNetCoreFunction`2.FunctionHandlerAsync(TREQUEST request, ILambdaContext lambdaContext)
   at lambda_method(Closure , Stream , Stream , LambdaContextInternal )
```

Your tip helped me! now it works. If anybody knows the explanation, I'll be happy to learn yet another new thing

**Comment by duaneking** [comment]
I just encountered this; a redeploy seems to have fixed it.  It looks like there is an error in the cloud front update when you do a serverless deploy that can lead to an invalid state on the lambda. No idea how or why, AWS was just broken. A redeploy of the exact same code fixed it, so it could not have been our code.

**Comment by nootle** [comment]
I was having the same issue when calling from API gateway to my lambda using a Mapped Template, to resolve this I ensure pathParameters and queryString paramaters are both declared even if the request does not have any.

FIXED:

{
  "httpMethod": "GET",
  "resource": "/lambda/resources",
  "path": "/lambda/resources",
  "pathParameters": {
  },
  "queryStringParameters": {
  },
  "headers": {  
  },
  "requestContext": {
    "domainName": "$context.domainName"
  }
}

ERRORS:

{
  "httpMethod": "GET",
  "resource": "/lambda/resources",
  "path": "/lambda/resources",
  "headers": {  
  },
  "requestContext": {
    "domainName": "$context.domainName"
  }
}

**Comment by duaneking** [comment]
Still getting these as random, its annoying and it inflates our bill because we have to pay for these exceptions being thrown by the AWS product itself when they show up in cloud watch.  Nothing seems to help ad t just randomly goes away, then comes back, even if we have not deployed.

**Comment by torbjokv** [comment]
In my case I had to change base class from **APIGatewayProxyFunction** to **APIGatewayHttpApiV2ProxyFunction**. Leaves this here in case anybody else have the same issue :-)

**Comment by jakobgoettrup** [comment]
getting this error after changing the default serverless project to use APIGatewayHttpApiV2ProxyFunction and JsonConvert.DeserializeObject<APIGatewayHttpApiV2ProxyRequest>(requestStr)

**Comment by torbjokv** [comment]
> getting this error after changing the default serverless web API project to use **APIGatewayHttpApiV2ProxyFunction** and `JsonConvert.DeserializeObject<APIGatewayHttpApiV2ProxyRequest>(requestStr)`
> 
> Steps to reproduce:
> 
> * 1 Create new project from template.
> * 2 Change to use **APIGatewayHttpApiV2ProxyFunction** in **LambdaEntryPoint**
> * 3 Change to use  **APIGatewayHttpApiV2ProxyRequest** in **ValuesControllerTests**
> * 4 Run test

Make sure your API Gateway integration have the correct payload format version set (API Gateway -> Integrations -> Manage integrations -> Payload format version)

**Comment by hasanrihawi** [comment]
Thanks @torbjokv your solution helped me out !!!

**Comment by harry-hathorn** [comment]
@torbjokv In my case i had to change from APIGatewayHttpApiV2ProxyFunction to APIGatewayProxyFunction

It seems like APIGatewayProxyFunction works for using a REST api gateway and APIGatewayHttpApiV2ProxyFunction works for HTTP api gateway.
